Refactored CI/CD and Docker Env

ysibirski opened the pull request:
This PR introduces major refactoring for CI/CD and Docker ENV

> ⚠️ 

- This PR introduces breaking changes for **compose-zkverify** project(no more _CONF_VAR_2. Entrypoint supports a list now)
- **Cache** per **MAIN** branch ONLY still due to current cache implementation on GH Actions level: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache
- Compose project needs to be updated once new image is built
- `--password-file` is going to be implemented in another PR
- Running CI locally tool `act` is being removed due to CI/CD workflow builds running in parallel vs sequential
- `staging/raw-chainspec.json` file is missing which is a requirement to run zkv-relay-docker-compose.yml project
- CI/CD build is being moved onto WarpBuild environment with unlimited cache and 50% off runners price vs GH Actions ones: https://www.warpbuild.com/products/ci-runners
- All build jobs can now be manually triggered as well

lander86 reviewed (approved):
LGTM . We'll have to fix --password-file logic in a later PR as just mentioned and another PR needs to follow this one to be sure to make compose repositories compatible.

la10736 commented on docker/dockerfiles/zkv-node.Dockerfile:
```diff
@@ -1,74 +1,75 @@
-  && apt-get -y autoremove \
-  && rm -rf /var/{lib/apt/lists/*,cache/apt/archives/*.deb} /tmp/*
-
-USER ${RUN_USER}
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+      aria2 \
+      ca-certificates \
+      curl \
+      jq \
+    && useradd -m -U -s /bin/bash -d "/${RUN_USER}" "${RUN_USER}" \
+    && mkdir -p /data "/${RUN_USER}/.local/share" \
+    && chown -R "${RUN_USER}:${RUN_USER}" /data "/${RUN_USER}" \
+    && ln -s /data "/${RUN_USER}/.local/share" \
+    && apt-get -y clean \
+    && apt-get -y autoclean \
+    && apt-get -y autoremove \
+    && rm -rf /var/{lib/apt/lists/*,cache/apt/archives/*.deb} /tmp/*
+
```
Maybe move all these installation before copy the binary from the other layer generate a more stable cache that can be reused more times (on developer host of course)

ysibirski replied:
Good point

la10736 commented on docker/dockerfiles/zkv-relay.Dockerfile:
```diff
@@ -1,76 +1,77 @@
-  && apt-get -y autoremove \
-  && rm -rf /var/{lib/apt/lists/*,cache/apt/archives/*.deb} /tmp/*
-
-USER ${RUN_USER}
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+      aria2 \
+      ca-certificates \
+      curl \
+      jq \
+    && useradd -m -U -s /bin/bash -d "/${RUN_USER}" "${RUN_USER}" \
+    && mkdir -p /data "/${RUN_USER}/.local/share" \
+    && chown -R "${RUN_USER}:${RUN_USER}" /data "/${RUN_USER}" \
+    && ln -s /data "/${RUN_USER}/.local/share" \
+    && apt-get -y clean \
+    && apt-get -y autoclean \
+    && apt-get -y autoremove \
+    && rm -rf /var/{lib/apt/lists/*,cache/apt/archives/*.deb} /tmp/*
```
also here... maybe is better to move them at line 48

la10736 commented on docker/scripts/entrypoint.sh:
```diff
@@ -10,16 +10,14 @@
 # Environment variables in the form `ZKV_CONF_*` are translated to command line arguments based on these rules:
 #
 # 1. `ZKV_CONF_` prefix is removed
-# 2. if both a trailing underscore (`_`) and number are present, they are removed
```
Why are you removing this behavior? we introduced it for every cases where we need to map array flags: flags that can be used more than once like `--bootnodes`.

ysibirski replied:
This is not being removed, rather being improved. With new implementation you no longer need to pass ONE value per VARIABLE if it supports list. Now you can do the following:

OLD
```
ZKV_CONF_LISTEN_ADDR="/ip4/0.0.0.0/tcp/${NODE_NET_P2P_PORT}"
ZKV_CONF_LISTEN_ADDR_2="/ip4/0.0.0.0/tcp/${NODE_NET_P2P_PORT_WS}/ws"
```

NEW
```
ZKV_CONF_LISTEN_ADDR="/ip4/0.0.0.0/tcp/${NODE_NET_P2P_PORT},/ip4/0.0.0.0/tcp/${NODE_NET_P2P_PORT_WS}/ws"
```

la10736 replied:
Ok, I got it. We chose the other solution because it also possible to use it when the arguments can contain also `,`. But it's ok for now.... if we'll meet some issue we'll change it later.

la10736 commented on docker/scripts/entrypoint.sh:
```diff
@@ -101,91 +107,95 @@ validate_and_download() {
-    fi
-    echo "  ${var_name}=${var_value} -> ${arg_name} ${arg_value}"
+# Loop through all arguments to check for --dev
+for arg in "$@"; do
+  if [ "${arg}" == "--dev" ]; then
+    log_green "INFO: '--dev' flag detected! Running in development mode."
+    # You can place any logic for development mode here
+    DEV_MODE="true"
```
Nice

la10736 commented on .github/workflows/CI-audit.yml:
```diff
@@ -8,32 +8,61 @@ on:
 jobs:
   audit:
-    runs-on: ubuntu-latest
+    runs-on: warp-ubuntu-latest-x64-2x
```
That's a really fast job. Maybe can still run on github. What about to have 2 variables: one for fast job and other for heavy jobs?

ysibirski replied:
We are running this on SMALLER instance anyway and paying 75% less then if we would run on GH Actions default runner.

la10736 commented on .github/workflows/CI-audit.yml:
```diff
@@ -8,32 +8,61 @@ on:
+    needs: [audit]
+    if: ${{ failure() && !cancelled() && github.event_name == 'schedule' }}
+    steps:
+      - name: Send Slack notification
+        uses: slackapi/slack-github-action@v1.25.0
+        with:
+          payload: |
```
What about to do our own slack action that use the generic one but with just 2 arguments: `header` and `text`?

la10736 replied:
I' don't know if it is possible to have an array for `text` in order to cover the multiple sections scenario.

ysibirski replied:
That is a good idea, but to be honest with you IF we go this path, we should deploy it as an action that can be shared by ALL our repos, therefore thinking on creating centralized GH Actions repo and then using those in our workflows since we are using this `SLACK-github-action` under each of the pipeline we wrote.

I would leave this improvement for future task since we probably want to think on how this should be implemented :shrug:

DanieleDiBenedetto commented on .github/workflows/CI-build.yml:
```diff
@@ -1,82 +1,84 @@
-        run: |
-          echo "USER_ID=$(id -u)" >> $GITHUB_ENV
-          echo "GRP_ID=$(id -g)" >> $GITHUB_ENV
+      - name: Upload zkv-node binary artifact
```
The binary is pretty big, are we sure we want to upload it ?

DanieleDiBenedetto replied:
Of course, but why uploading them ? Can't we save them locally somewhere ?

ysibirski replied:
That is the only way to GET the artifact using a DIFF workflow/job. Artifacts space is free but has some limits like single artifact file can not exceed 2 GB and etc. In our case we use default compression level = 6 and its total size is 21MB

DanieleDiBenedetto commented on .github/workflows/CI-build.yml:
```diff
@@ -1,82 +1,84 @@
+          lld_install: 'yes'
 
-      - name: Upload relay chain runtime artifact
+      - name: Upload relay and paratest build binaries
```
Same comment as before, too big to upload them ?

la10736 replied:
We need them for zombienet. Don't it?

DanieleDiBenedetto replied:
Of course, but why uploading them ? Can't we save them locally somewhere ?

ysibirski replied:
Same answer as before. It is ok to upload them as artifacts, stored for 1 day with default compression level 6

![image](https://github.com/user-attachments/assets/532b24ac-d0bb-4d5a-9391-daeacb38a733)

DanieleDiBenedetto commented on .github/workflows/CI-build.yml:
```diff
@@ -1,82 +1,84 @@
+          cmake_install: 'yes'
+          lld_install: 'yes'
+
+      - name: Upload try-runtime wasm artifact
```
No need for this

la10736 replied:
I guess that here we already have a lot of stuff cached ... Rebuilt it on `CI-try-runtime` means recompile them. Anyway it makes the sequential... but the just try step without compile is really fast (2 minutes) so maybe this approach is a little better

ysibirski replied:
Adjusting

la10736 commented on ci/docker.sh:
It's not mater of your PR, but now also this name is misleading: it's just about publish image and extract wasm... not, like the name suggest, building image.

ysibirski replied:
Renaming to publish-docker-image instead

DanieleDiBenedetto commented on .github/workflows/CI-test.yml:
```diff
@@ -4,83 +4,80 @@ run-name: "Workflow CI/CD Steps: build, unit and integration testing"
+      - name: Cargo unit test WITHOUT relay feature
+        uses: ./.github/actions/cargo-cmd
+        with:
+          command: "cargo test --lib --bins --all-features --no-fail-fast --release 2>&1 | tee unit_test_no_relay_output.txt"
```
I think you meant to pass `--all-features` ?

la10736 replied:
No, I meant exactly what I wrote, we're just testing `zkv-runtime` crate *without* the `relay` feature and with `runtime-feature`. When we tested all crates with all features the runtime was tested with also the `relay` features activated, but when we use the `zkv-runtime` in solo chain the feature is disabled and the runtime is really different.

ysibirski commented on .github/actions/cargo-cmd/action.yml:
```diff
@@ -0,0 +1,119 @@
+        docker compose -f ${GITHUB_WORKSPACE}/ci/docker-compose.yml run --rm cargo-container ${{ inputs.command }}
+
+    # Clear cache key for default_branch triggered by merge into default_branch event ONLY
+    - name: Clear cache before saving new version
+      id: clear_cache
+      if: ${{ inputs.use_cache == 'yes' && github.ref == format('refs/heads/{0}', github.event.repository.default_branch) }}
+      uses: WarpBuilds/cache@v1
+      with:
+        path: |
+          ~/.cargo/bin/
+          ~/.cargo/registry/index/
+          ~/.cargo/registry/cache/
+          ~/.cargo/git/db/
+          target/
+          deps/
+        key: ${{ inputs.cache_key }}-${{ github.event.repository.default_branch }}
+        delete-cache: true
```
Yes, unfortunately GH Actions cache implementation enforce CACHE_KEY to be immutable, therefore if we want to overwrite the existing cache key coming from the same branch we have to DELETE it first and WRITE again.

ysibirski commented on .github/workflows/CI-build-test-publish.yml:
Agreed, I was thinking about changing this as well. Will take care of it

ysibirski commented on .github/actions/cargo-cmd/action.yml:
Addressed

ysibirski commented on .gitignore:
```diff
@@ -4,6 +4,7 @@ target/
 *.iml
 .secrets
 .vscode
+actionlint.yaml
```
I think it is a very good idea for all of US working on the pipeline to use linter. I have discovered it and it looks/works very good. Even, if we include actionlint.yaml under gitignore it is not enforcing its usage, but removes the headache of someone who will actually use it.

I am indifferent on this one. @la10736 let me know what do you prefer. I can always add this file and delete it all the time I am working on the pipeline, but thought keep it here will not hurt :shrug:

ysibirski replied:
Adding

ysibirski commented on docker/dockerfiles/zkv-node.Dockerfile:
```diff
@@ -1,74 +1,75 @@
 WORKDIR /usr/src/node
 COPY . .
 
-RUN echo "SUBSTRATE_CLI_GIT_COMMIT_HASH=`git rev-parse --short=11 HEAD`" >> .docker.env
-RUN . ./.docker.env \
-  && cargo build -p mainchain --profile ${PROFILE} --features "${FEATURES}"
```
I think I was not sure how do we use this `SUBSTRATE_CLI_GIT_COMMIT_HASH` var?  Is it actually being used for `cargo build` command? Tell me more about this please.

la10736 replied:
Ok, you're right!!! When I added this code to solve the issue I did 2 changes

- this line
- removed from docker ignore file the rows about ignore `.git` and `.gitignore` 

Just the second was enough to solve the issue!!!

The code used in `build.rs` script via`generate_cargo_keys()` https://docs.rs/substrate-build-script-utils/latest/src/substrate_build_script_utils/version.rs.html#21-50 do exactly this one if the the variable isn't defined.

Thanks again

ysibirski commented:
@la10736 and @DanieleDiBenedetto please read all my replies and take a look at my latest commit where I addressed all the comments I had not further questions or explanation about

la10736 commented on ci/setup_env.sh:
```diff
@@ -26,21 +26,22 @@ fi
-    log_warn "WARNING: there are ZERO gpg keys to import. Please check if MAINTAINERS_KEYS variable(s) is(are) set correctly. The build is not going to be released ..."
-    export IS_A_RELEASE="false"
+  if [ "${#keys[@]}" -eq 0 ]; then
+    log_warn "WARNING: there are ZERO gpg keys to import. Please check if 'MAINTAINERS_KEYS' variable is set correctly. The build is not going to be released ..."
```
Ok,we'll do it in a next PR

la10736 reviewed (approved):
LGTM

Could you run it on the other repo before merge it?

ysibirski commented on docker/README.md:
````diff
@@ -17,7 +17,7 @@ Start a _**solo**_ chain with 2 validators (Alice and Bob) and a simple client n
 ```
 
-Start a _**relay**_ chain with a test _**parachain**_ with:
+**TO BE SUPPORTED:** Start a _**relay**_ chain with a test _**parachain**_ with:
````
Yes, it requires spec file under ../../staging dir and there is nothing there, but perhaps it will be worth mentioning that file needs to be placed there?
